### PR TITLE
Create depletion functions

### PR DESCRIPTION
**Pull request by abachma2**
This PR adds a class to call the ``IndependentOperator`` class in Open MC to perform depletion and functions related to calling this class and converting the output compositions into a file format readable by Cyclus. This PR also adds a test suite for unit tests for these functions, as well as file needed to run the test suite. All tests pass on my local machine, I have an issue (#4) to get CI set up for tests. 

I expect that these functions will be updated/augmented once the coupling to the Cyclus archetype begins, but these functions provide a basis for what I believe will be needed. I am open to comments/suggestions on style and content.

### CODE REVIEW

File: openmcyclus/depletion.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            geometry=str(
+                self.path +
+                "geometry.xml"),
+            materials=str(
+                self.path +
+                "materials.xml"),
+            settings=str(
+                self.path +
+                "materials.xml"))
+        return model
```
Comment (yardasol) [inline openmcyclus/depletion.py:59]: This allows optional support for `tallies.xml`
```suggestion
        model_kwargs = {"geometry": (self.path / "geometry.xml"),
                        "materials": (self.path / "materials.xml"),
                        "settings": (self.path / "settings.xml")}
        tallies_path = self.path / "tallies.xml"
        if tallies_path.exists():
            model_kwargs["tallies"] = tallies_path
        model = openmc.Model.from_xml(**model_kwargs)
        return model
```

File: openmcyclus/depletion.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+        Parameters:
+        -----------
+
+        Returns:
+        --------
+        microxs: object
+            microscopic cross section data
+        '''
+        microxs = od.MicroXS.from_csv(str(self.path + "micro_xs.csv"))
```
Comment (yardasol) [inline openmcyclus/depletion.py:75]:
```suggestion
        microxs = od.MicroXS.from_csv(self.path / "micro_xs.csv")
```

File: openmcyclus/depletion.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+        Outputs:
+        --------
+        {self.path}/{self.prototype}_fuel.xml : XML file
+            File containing the depleted material composition in the
+            required format for Cyclus. The path and prototype names are read
+            in from the class instantiation. Path will need to
+            be relative to Cyclus input file location
+        '''
+        results = od.Results.from_hdf5(str(self.path + "depletion_results.h5"))
```
Comment (yardasol) [inline openmcyclus/depletion.py:120]: `Results.from_hdf5` is being depreciated. 
```suggestion
        results = od.Results(self.path / "depletion_results.h5")
```

File: tests/unit_tests/test_depletion.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        self.deplete = Depletion(
+            "tests/", 'Reactor', "chain_endfb71_pwr.xml", 10, 100)
+
+    def test_read_model(self):
+        '''
+        Test for when the files are found
+        '''
+        model = self.deplete.read_model()
+        obs = isinstance(model, openmc.model.model.Model)
+        assert obs
```
Comment (yardasol) [inline tests/unit_tests/test_depletion.py:26]:
```suggestion
        assert isinstance(model, openmc.model.model.Model)
```

File: tests/unit_tests/test_depletion.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    def test_read_microxs(self):
+        microxs = self.deplete.read_microxs()
+        obs = isinstance(microxs, pd.DataFrame)
+        assert obs
```
Comment (yardasol) [inline tests/unit_tests/test_depletion.py:31]:
```suggestion
        assert isinstance(microxs, pd.DataFrame)
```

File: tests/unit_tests/test_depletion.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        assert microxs.index.name == 'nuclide'
+        assert microxs.columns.values.tolist() == [
+            '(n,gamma)', '(n,2n)', '(n,p)', '(n,a)', '(n,3n)', '(n,4n)', 'fission']
+
+    def test_run_depletion(self):
+        self.deplete.run_depletion()
+        obs = os.path.exists(str(self.deplete.path + "depletion_results.h5"))
+        assert obs
```
Comment (yardasol) [inline tests/unit_tests/test_depletion.py:39]:
```suggestion
        assert (self.deplete.path / "depletion_results.h5").exists()
```

File: tests/unit_tests/test_depletion.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    def test_create_recipe(self):
+        self.deplete.run_depletion()  # make sure database is present
+        self.deplete.create_recipe()
+        os.system("rm " + self.deplete.path + "depltion_results.h5")
```
Comment (yardasol) [inline tests/unit_tests/test_depletion.py:44]:
```suggestion
       (self.deplete.path / "depltion_results.h5").unlink()
```

File: tests/unit_tests/test_depletion.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
+import numpy as np
+import pytest
+import xml.etree.ElementTree as ET
+import unittest
+import openmc
+import pandas as pd
+from openmcyclus.depletion import Depletion
+
+import os
```
Comment (yardasol) [inline tests/unit_tests/test_depletion.py:9]:
```suggestion
import pathlib
```

**Review by yardasol — CHANGES_REQUESTED** [review]
Hi @abachma2, this looks fantastic! I have one big overall comments:
I'd highly recommend using [`pathlib`](https://docs.python.org/3/library/pathlib.html#correspondence-to-tools-in-the-os-module) instead of OS, it's a more pythonic way of dealing with filesystems and in my experience is slightly less clunky. `pathlib` can't do _everything_ that `os` can, but for this use case it's more appropriate.

All my other comments and suggestions are below.

File: openmcyclus/depletion.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        micro_xs = self.read_microxs()
+        ind_op = od.IndependentOperator(model.materials, micro_xs,
+                                        str(self.path + self.chain_file))
+        integrator = od.PredictorIntegrator(ind_op, np.ones(
+            self.timesteps), power=self.power, timestep_units='d')
+        integrator.integrate()
+        os.system(
+            'mv ./depletion_results.h5 ' +
+            self.path +
+            "depletion_results.h5")
```
[the remark this answers is not in the document]
Comment (abachma2) [reply openmcyclus/depletion.py:101]: I looked for a way to do that within OpenMC. Guess I missed that one. Thanks!

File: README.md [hunk 10]
```diff
@@ -4,3 +4,21 @@ Repository of [Cyclus](https://fuelcycle.org/) archetypes to couple Cyclus with
...
+To run tests: ``pytest`` from the main directory.
+
+### Dependencies
+You will need to have [Cyclus](www.github.com/cyclus/cyclus), [OpenMC](https://docs.openmc.org).
+and their required dependencies. Install both of these in the same conda environment.
+
+### Running
+This archetype assumes that you have a defined reactor model in OpenMC (``.xml``) 
+files and the 
+required microscopic cross sections (``.xml`` file)for the model. The cross
```
[the remark this answers is not in the document]
Comment (abachma2) [reply README.md:16]: I mean the ``.csv`` file for the cross sections, like the ``micro_xs.csv`` file that I added to the repo. I am trying to reduce the number of things OpenMC has to do, and I feel like if we tried to use the ``openmc.deplete.MicroXS.from_model()`` method then that would calculate the cross sections each time you call the depletion solver, which I don't want to do. Please correct me if I am incorrect in my understanding.

File: openmcyclus/depletion.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+            name of prototype undergoing depletion
+        chain_file: str
+            name of file with decay chain data. Just the file name is
+            required, it is assumed that the file is in the same
+            location as the other files for the OpenMC model
+        timesteps: int
+            number of time steps to perform irradiation. It is assumed that
+            the number given is in units of months.
+        power: int
+            power output of the reactor, assumed in MWe.
```
[the remark this answers is not in the document]
Comment (abachma2) [reply openmcyclus/depletion.py:29]: This is a really interesting question. Cyclus assumes the power is in MWe. So I feel like there are a few ways we can handle this:
- Convert MWe to MWth in the archetype, assuming a constant conversion
- Convert MWe to MWth in a method in this class, assuming a constant conversion
- Convert MWe to MWth in the archetype with a user defined conversion 
- Convert MWe to MWth in a method in this class with a user defined conversion
Thoughts? I'm okay with moving this into an issue and discussing later if you feel this isn't vital to this PR.

File: openmcyclus/depletion.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (abachma2) [reply openmcyclus/depletion.py]: I changed the code to what you suggested and when I run the tests I get ``TypeError: IndependentOperator.__init__() got an unexpected keyword argument 'output_dir'``. Any thoughts?

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (yardasol) [reply README.md]: Gotcha. In that case you should just change the ``.xml`` to a ``.csv`` in the docstring.

You are correct, we do not want to call `openmc.deplete.MicroXS.from_model()` during runtime, that goes against the point of using `IndependentOperator`.

File: openmcyclus/depletion.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (yardasol) [reply openmcyclus/depletion.py]: My bad, `output_dir` is not a parameter for `IndependentOperator.__init__()`, it's an attribute that you'll need to set manually. So after creating `ind_op`, you should write
```python
ind_op.output_dir = self.path
```

File: openmcyclus/depletion.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (yardasol) [reply openmcyclus/depletion.py:32]: I think the fourth option would be easiest to do, and you can have a default conversion factor of 3, which will get you in the ballpark I think.

**Comment by abachma2** [comment]
I have made the remaining changes requested, and the tests all pass still.

**Comment by abachma2** [comment]
@yardasol bump

File: openmcyclus/depletion.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+from xml.dom import minidom
+import pathlib
+
+
+class Depletion(object):
+    def __init__(self, path:str, prototype:str, chain_file:str, 
+    timesteps:int, power:float, conversion_factor:float=3):
+        '''
+        Class to hold objects related to calling 
+        ~:class:`~openmc.deplete.IndependentOperator`
```
Comment (yardasol) [inline openmcyclus/depletion.py:13]:
```suggestion
        :class:`~openmc.deplete.IndependentOperator`
```

**Review by yardasol — APPROVED** [review]
Apologies for the long wait on this @abachma2. Great work on these changes. Just one quick change and then I'll merge

**Pull request merged**